torch/csrc/cuda/comm.cpp

```diff
@@ -144,8 +145,9 @@ tensor_list2d broadcast_coalesced(TensorList tensors, IntArrayRef devices, size_
         for (auto & t : utils::unflatten_dense_tensors(results[i], chunk.tensors)) {
           // See NOTE [ Version Counter in comm.*_coalesced ]
           AT_ASSERT(t.is_variable());
-          Variable var = t;
-          device_outputs.push_back(make_variable(var.tensor_data(), false));
+          Variable var = t.variable_data();
+          var.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
+          device_outputs.push_back(var);
```

Comment:
We set `var`'s `allow_tensor_metadata_change` to true before putting it into `device_outputs`, because we want to allow the following use case:
```python
class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
        self.rnn = torch.nn.LSTM(300, 1024, 1, batch_first=True, bidirectional=True)

    def forward(self, x):
        self.rnn.flatten_parameters() # this internally calls `torch._cudnn_rnn_flatten_weight(...)`, which needs to change storage of `rnn`'s weights in-place 
        return self.rnn(x)


with torch.no_grad():
    model_dp = torch.nn.DataParallel(TestModule().to(0))
    model_dp(torch.rand(2, 4, 300).to(0))
```

Reply:
is this the only place where this comes up?  Why does data parallel need to flatten the parameters in place?

Reply:
This problem is actually not specific to data parallel - https://github.com/pytorch/pytorch/pull/22172/files#diff-13b395c368197e2cac32474f78c76ae4R159 has the same problem too. The main issue is that `flatten_parameters` in RNN internally calls `torch._cudnn_rnn_flatten_weight(...)`, which needs to change storage of RNN's weights in-place. Therefore, we need to set the `allow_tensor_metadata_change` flag to true here in order to make the example work.

Reply:
I don't think this really answered the question, the question was: "Why does data parallel need to flatten the parameters in place?"

The response was: "which needs to change storage of RNN's weights in-place."

Why does it need to change storage of RNN's weights in-place?

Reply:
Flattening the parameters of an RNN before calling forward on it is actually a common use case, and we have warnings that encourage people to do so:
```python
m = torch.nn.LSTM(300, 1024, 1).to('cuda')
m(torch.rand(2, 4, 300).to('cuda'))

# Output:
# ../aten/src/ATen/native/cudnn/RNN.cpp:1266: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().
#  AT_WARN(WEIGHT_FORMAT_WARN);
```
When we flatten the parameters of an RNN, we need to allocate a new chunk of contiguous CUDA memory, and then point the RNN's parameters to the new memory. Therefore we have to change storage of RNN's weights in-place.

This problem is not specific to DataParallel, but in our test suite we have code that runs RNN (and flattens its parameters) in a DataParallel setting, and to support that use case we need to set the `allow_tensor_metadata_change` flag to true here.